src/soc/ibm/power9/powerbus.c

```diff
@@ -0,0 +1,244 @@
+	} else if ((2 * freq_ceiling) >= pb_freq) {
+		ceiling_ratio = FABRIC_CORE_CEILING_RATIO_RATIO_2_8;
+	} else {
+		printk(BIOS_ERR, "Unsupported core ceiling/PB frequency ratio = (%d/%d)\n",
+			freq_ceiling, pb_freq);
+		return false;
+	}
+
+	cfg->core_floor_ratio = floor_ratio;
+	cfg->core_ceiling_ratio = ceiling_ratio;
+	return true;
+}
+
+static void config_guardband_epsilon(uint8_t gb_percentage, uint32_t *target_value)
+{
+	uint32_t delta = (*target_value * gb_percentage) / 100;
+	delta += ((*target_value * gb_percentage) % 100) ? 1 : 0;
+
+	/* Clamp to maximum value if necessary */
+	if (delta > (EPSILON_MAX_VALUE - *target_value)) {
+		printk(BIOS_DEBUG, "Guardband application generated out-of-range target value,"
+		       " clamping to maximum value!\n");
+		*target_value = EPSILON_MAX_VALUE;
+	} else {
+		*target_value += delta;
+	}
+}
+
+static void dump_epsilons(struct powerbus_cfg *cfg)
+{
+	uint32_t i;
+
+	for (i = 0; i < NUM_EPSILON_READ_TIERS; i++)
+		printk(BIOS_DEBUG, " R_T[%d] = %d\n", i, cfg->eps_r[i]);
+
+	for (i = 0; i < NUM_EPSILON_WRITE_TIERS; i++)
+		printk(BIOS_DEBUG, " W_T[%d] = %d\n", i, cfg->eps_w[i]);
+}
+
+static void calculate_epsilons(struct powerbus_cfg *cfg)
+{
+	const enum FABRIC_CORE_FLOOR_RATIO floor_ratio = cfg->core_floor_ratio;
+	const enum FABRIC_CORE_CEILING_RATIO ceiling_ratio = cfg->core_ceiling_ratio;
+	const uint32_t pb_freq = cfg->fabric_freq;
+	const uint32_t freq_ceiling = cfg->freq_core_ceiling;
+
+	uint32_t *eps_r = cfg->eps_r;
+	uint32_t *eps_w = cfg->eps_w;
+
+	uint64_t scratch_reg6 = read_scom(MBOX_SCRATCH_REG1 + 5);
+	/* ATTR_PROC_FABRIC_PUMP_MODE, it's either node or group pump mode */
+	bool node_pump_mode = !(scratch_reg6 & MBOX_SCRATCH_REG6_GROUP_PUMP_MODE);
+
+	/* Assuming that ATTR_PROC_EPS_TABLE_TYPE = EPS_TYPE_LE in talos.xml is always correct*/
+
+	eps_r[0] = EPSILON_R_T0_LE[floor_ratio];
+
+	if (node_pump_mode)
+		eps_r[1] = EPSILON_R_T1_LE[floor_ratio];
+	else
+		eps_r[1] = EPSILON_R_T0_LE[floor_ratio];
+
+	eps_r[2] = EPSILON_R_T2_LE[floor_ratio];
+
+	eps_w[0] = EPSILON_W_T0_LE[floor_ratio];
+	eps_w[1] = EPSILON_W_T1_LE[floor_ratio];
+
+	/* Dump base epsilon values */
+	printk(BIOS_DEBUG, "Base epsilon values read from table:\n");
+	dump_epsilons(cfg);
+
+	/* Scale base epsilon values if core is running 2x nest frequency */
+	if (ceiling_ratio == FABRIC_CORE_CEILING_RATIO_RATIO_8_8) {
+		uint32_t i;
+
+		uint8_t scale_percentage = 100 * freq_ceiling / (2 * pb_freq);
+		if (scale_percentage < 100)
+			die("scale_percentage is too small!");
+		scale_percentage -= 100;
+
+		printk(BIOS_DEBUG, "Scaling based on ceiling frequency\n");
+
+		/* scale/apply guardband read epsilons */
+		for (i = 0; i < NUM_EPSILON_READ_TIERS; i++)
+			config_guardband_epsilon(scale_percentage, &eps_r[i]);
+
+		/* Scale write epsilons */
+		for (i = 0; i < NUM_EPSILON_WRITE_TIERS; i++)
+			config_guardband_epsilon(scale_percentage, &eps_w[i]);
+	}
+
```

Comment:
Hostboot in this place scales epsilons even for ratios different than 8/8:
https://github.com/open-power/hostboot/blob/master/src/import/chips/p9/procedures/hwp/nest/p9_fbc_eff_config.C#L300
We should do the same, at least to make it easier to compare register values.

Reply:
The code looked weird in this place, scaling must have been removed by accident while extracting `dump_epsilons()`. Scaled values:
```
Scaled epsilon values based on +20 percent guardband:
 R_T[0] = 12
 R_T[1] = 12
 R_T[2] = 95
 W_T[0] = 0
 W_T[1] = 26
```